### PR TITLE
[FLINK-13435] Remove ShuffleDescriptor.ReleaseType and make release semantics fixed per partition type

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

In a long term we do not need auto-release semantics for blocking (persistent) partition. We expect them always to be released externally by JM and assume they can be consumed multiple times.

The pipelined partitions have always only one consumer and one consumption attempt. Afterwards they can be always released automatically.

`ShuffleDescriptor#ReleaseType` was introduced to make release semantics more flexible but it is not needed in a long term.

`FORCE_PARTITION_RELEASE_ON_CONSUMPTION` was introduced as a safety net to be able to fallback to 1.8 behaviour without the partition tracker and JM taking care about blocking partition release. We can make this option specific for `NettyShuffleEnvironment` which was the only existing shuffle service before. If it is activated then the blocking partition is also auto-released on a consumption attempt as it was before. The fine-grained recovery will just not find the partition after the job restart in this case and will restart the producer.

## Brief change log

  - Remove `ShuffleDescriptor#ReleaseType` and assume predefined release semantics for certain partition type
  - Adjust test and other involved code
  - Move `FORCE_PARTITION_RELEASE_ON_CONSUMPTION` from `JobManagerOptions` to `NettyShuffleEnvironmentOptions`
  - codestyle fixes

## Verifying this change

existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NettyShuffleEnvironmentConfiguration.java [hunk 1]
```diff
@@ -29,8 +29,8 @@
 import org.apache.flink.runtime.io.network.netty.NettyConfig;
 import org.apache.flink.runtime.io.network.partition.BoundedBlockingSubpartitionType;
 import org.apache.flink.runtime.util.ConfigurationParserUtils;
-
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NettyShuffleEnvironmentConfiguration.java]: This change should be a separate hotfix

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NettyShuffleEnvironmentConfiguration.java [hunk 2]
```diff
@@ -73,6 +73,8 @@
 
 	private final BoundedBlockingSubpartitionType blockingSubpartitionType;
 
+	private final boolean forcePartitionReleaseOnConsumption;
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NettyShuffleEnvironmentConfiguration.java]: Not sure to also cover this new field in following `hashCode`, `toString` and `equals` methods.

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java [hunk 3]
```diff
@@ -61,17 +60,14 @@
...
- *     {@link ShuffleMaster#releasePartitionExternally(ShuffleDescriptor)} and {@link ShuffleEnvironment#releasePartitionsLocally(Collection)},
- *     if it occupies any producer local resources ({@link ShuffleDescriptor#storesLocalResourcesOn()}),
- *     are called outside of the producer thread, e.g. to manage the lifecycle of BLOCKING result partitions
- *     which can outlive their producers. The partition has to support the {@link ReleaseType#MANUAL} in
- *     {@link ShuffleDescriptor#getSupportedReleaseTypes()}.</li>
+ *     <li>for bounded pipelined partitions if the production is done ({@link ResultPartitionWriter#finish()} and
+ *     {@link ResultPartitionWriter#close()} are called) and there was a detected consumption attempt.
+ *     Only one consumption attempt is ever expected for the pipelined partition at the moment so
+ *     it cqn be released afterwards.
+ *     <li>{@link ShuffleMaster#releasePartitionExternally(ShuffleDescriptor)} and {@link ShuffleEnvironment#releasePartitionsLocally(Collection)},
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: There might some confusing here. `ShuffleMaster#releasePartitionExternally` is mainly used for releasing external resources outside of local producer. So it is not very inconsistent with `if it occupies any producer local resources`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java [hunk 4]
```diff
@@ -61,17 +60,14 @@
...
+ *     if it occupies any producer local resources ({@link ShuffleDescriptor#storesLocalResourcesOn()}), are called
+ *     outside of the producer thread, e.g. to manage the lifecycle of BLOCKING or pipelined bounded result partitions
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: also lowercase for `BLOCKING` consistent with pipelined

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java [hunk 5]
```diff
@@ -61,17 +60,14 @@
...
- *     <li>if {@link ResultPartitionDeploymentDescriptor#isReleasedOnConsumption()} is {@code false} and
- *     {@link ShuffleMaster#releasePartitionExternally(ShuffleDescriptor)} and {@link ShuffleEnvironment#releasePartitionsLocally(Collection)},
- *     if it occupies any producer local resources ({@link ShuffleDescriptor#storesLocalResourcesOn()}),
- *     are called outside of the producer thread, e.g. to manage the lifecycle of BLOCKING result partitions
- *     which can outlive their producers. The partition has to support the {@link ReleaseType#MANUAL} in
- *     {@link ShuffleDescriptor#getSupportedReleaseTypes()}.</li>
+ *     <li>for bounded pipelined partitions if the production is done ({@link ResultPartitionWriter#finish()} and
+ *     {@link ResultPartitionWriter#close()} are called) and there was a detected consumption attempt.
+ *     Only one consumption attempt is ever expected for the pipelined partition at the moment so
+ *     it cqn be released afterwards.
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: typo cqn

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java [hunk 6]
```diff
@@ -61,17 +60,14 @@
...
- *     The actual release can take some time depending on implementation details,
- *     e.g. if the `end of consumption' confirmation from the consumer is being awaited implicitly.
- *     The partition has to support the {@link ReleaseType#AUTO} in {@link ShuffleDescriptor#getSupportedReleaseTypes()}.</li>
- *     <li>if {@link ResultPartitionDeploymentDescriptor#isReleasedOnConsumption()} is {@code false} and
- *     {@link ShuffleMaster#releasePartitionExternally(ShuffleDescriptor)} and {@link ShuffleEnvironment#releasePartitionsLocally(Collection)},
- *     if it occupies any producer local resources ({@link ShuffleDescriptor#storesLocalResourcesOn()}),
- *     are called outside of the producer thread, e.g. to manage the lifecycle of BLOCKING result partitions
- *     which can outlive their producers. The partition has to support the {@link ReleaseType#MANUAL} in
- *     {@link ShuffleDescriptor#getSupportedReleaseTypes()}.</li>
+ *     <li>for bounded pipelined partitions if the production is done ({@link ResultPartitionWriter#finish()} and
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: I am a bit confused here. From the following comments of `Only one consumption attempt is ever expected for the pipelined partition`, I guess the `bounded pipelined partitions` is only meaning current pipelined partitions. Then for pipelined partition we do not need the production done to trigger the consumption attempt.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PartitionTestUtils.java [hunk 7]
```diff
@@ -60,9 +57,9 @@ public static ResultPartition createPartition(ResultPartitionType type, FileChan
 	}
 
 	public static ResultPartition createPartition(
-			NettyShuffleEnvironment environment,
-			ResultPartitionType partitionType,
-			int numChannels) {
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PartitionTestUtils.java]: irrelevant changes.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PartitionTestUtils.java [hunk 8]
```diff
@@ -72,10 +69,10 @@ public static ResultPartition createPartition(
 	}
 
 	public static ResultPartition createPartition(
-			NettyShuffleEnvironment environment,
-			FileChannelManager channelManager,
-			ResultPartitionType partitionType,
-			int numChannels) {
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PartitionTestUtils.java]: ditto: irrelevant changes

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java [hunk 9]
```diff
@@ -139,10 +132,10 @@ private void createSubpartitions(
 			ResultPartitionType type,
 			BoundedBlockingSubpartitionType blockingSubpartitionType,
 			ResultSubpartition[] subpartitions) {
-
 		// Create the subpartitions.
 		switch (type) {
 			case BLOCKING:
+			case BLOCKING_PERSISTENT:
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java]: I think we might need a uint test for covering this change.

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleDescriptor.java [hunk 10]
```diff
@@ -41,14 +41,14 @@
 	 * that the producer of the partition (consumer input channel) has not been scheduled
 	 * and its location and other relevant data is yet to be defined.
 	 * To proceed with the consumer deployment, currently unknown input channels have to be
-	 * marked with placeholders which are special implementation of {@link ShuffleDescriptor}:
+	 * marked with placeholders which are special implementation of shuffle descriptor:
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleDescriptor.java]: implementation -> implementations?

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleDescriptor.java [hunk 11]
```diff
@@ -41,14 +41,14 @@
...
 	 * {@link UnknownShuffleDescriptor}.
 	 *
 	 * <p>Note: this method is not supposed to be overridden in concrete shuffle implementation.
 	 * The only class where it returns {@code true} is {@link UnknownShuffleDescriptor}.
 	 *
 	 * @return whether the partition producer has been ever deployed and
-	 * the corresponding {@link ShuffleDescriptor} is obtained from the {@link ShuffleMaster} implementation.
+	 * the corresponding shuffle descriptor is obtained from the {@link ShuffleMaster} implementation.
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleDescriptor.java]: Are there any concerns for this change?

**Comment by zhijiangW** [comment]
@azagrebin I have finished the whole review and left some nit comments besides the above concerns.

**Comment by azagrebin** [comment]
Thanks for the review @zhijiangW ! I will go through the smaller comments. For general concerns:

- I think this PR should fit to the overall change we plan with #9250 and FLINK-13476. In #9250 we basically stop using explicitly any notion of 'release on consumption' and assume that blocking partitions always have to be released by JM/TM. FLINK-13476 should also fix it for pipelined bounded until we couple Task finished state with the consumption confirmation attempt, basically with the partition state. Not sure, there are other ways for the direct release internally in netty shuffle service.

- FLINK-13478 and FLINK-13493 seem to be independent issues or do I miss something?

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java (hunk 6 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: I will rephrase it a bit, I meant that basically only bounded pipelined can be ever done.

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleDescriptor.java (hunk 11 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleDescriptor.java]: there are some check style warnings in Idea IDE to not use self links

**Comment by azagrebin** [comment]
Thanks for the review @zhijiangW ! I have addressed the comments

**Comment by zhijiangW** [comment]
Thanks for the updates @azagrebin !

Almost LGTM on my side, only confirm one thing before merging:
I thought `FORCE_PARTITION_RELEASE_ON_CONSUMPTION` only suitable for pipelined partitions. If we make this option true for blocking partitions, then the partition release should be still controlled by JM always. If so, do we need to avoid generate `ReleaseOnConsumptionResultPartition` for blocking partitions in `ResultPartitionFactory`? In current implementation, the `ReleaseOnConsumptionResultPartition` would be constructed if `forcePartitionReleaseOnConsumption || !type.isBlocking()`.

I wonder the blocking partition release controlled by `ReleaseOnConsumptionResultPartition` might cause some problem because of FLINK-13493.

**Comment by azagrebin** [comment]
Thanks for the review @zhijiangW !

I mentioned the explanation in FLINK-13493. `forcePartitionReleaseOnConsumption` is just a safety net to fallback to the old 1.8 behaviour and should not be used for 1.9 in general. JM does not use this option and always tries to take care about release (as you mentioned it should be always the case) so `forcePartitionReleaseOnConsumption` will just enforce/speed up release in certain cases and degrade fine-grained recovery. In future, we can remove `forcePartitionReleaseOnConsumption` at all and rely only on JM.

**Comment by zhijiangW** [comment]
Thanks for the explanation @azagrebin !

I understand it, wish the temporary `forcePartitionReleaseOnConsumption` removed soon to avoid unnecessary concern here. :)
Will merge after travis green.

**Comment by zhijiangW** [comment]
Unrelated python component failure in travis, merging.

**Pull request merged**